File: ga/gpt3_5_result_2024_6/result/multi_thread_basic_gpt/9_multi_thread_basic.c
use std::thread;
use std::sync::{Mutex, Arc};

// Global shared variable representing account balance
static mut ACCOUNT_BALANCE: i32 = 1000000;
static MUTEX: Mutex<()> = Mutex::new(());

fn withdraw() {
    let amount = 500000;
    for _ in 0..amount {
        unsafe {
            let _lock = MUTEX.lock().unwrap();
            ACCOUNT_BALANCE -= 1;
        }
    }
}

fn main() {
    let mutex = Arc::new(Mutex::new(()));

    let mutex1 = Arc::clone(&mutex);
    let handle1 = thread::spawn(move || {
        withdraw();
    });

    let mutex2 = Arc::clone(&mutex);
    let handle2 = thread::spawn(move || {
        withdraw();
    });

    handle1.join().unwrap();
    handle2.join().unwrap();

    unsafe {
        println!("Final account balance: {}", ACCOUNT_BALANCE);
    }
}

File: ga/gpt3_5_result_2024_6/result/multi_thread_basic_cpp_gpt/9_multi_thread_basic.rs
use std::sync::{Arc, Mutex};
use std::thread;

fn withdraw(account_balance: &Arc<Mutex<i32>>) {
    let mut balance = account_balance.lock().unwrap();
    for _ in 0..500000 {
        *balance -= 1;
    }
}

fn main() {
    let account_balance = Arc::new(Mutex::new(1000000));

    let account_balance_clone = Arc::clone(&account_balance);
    let t1 = thread::spawn(move || {
        withdraw(&account_balance_clone);
    });

    let account_balance_clone = Arc::clone(&account_balance);
    let t2 = thread::spawn(move || {
        withdraw(&account_balance_clone);
    });

    t1.join().unwrap();
    t2.join().unwrap();

    let final_balance = account_balance.lock().unwrap();
    println!("Final account balance: {}", *final_balance);
}